services/src/main/java/org/keycloak/protocol/oidc/mappers/AcrProtocolMapper.java

```diff
@@ -0,0 +1,132 @@
+
+    static {
+        OIDCAttributeMapperHelper.addIncludeInTokensConfig(configProperties, AcrProtocolMapper.class);
+    }
+
+    public static final String PROVIDER_ID = "oidc-acr-mapper";
+
+
+    public List<ProviderConfigProperty> getConfigProperties() {
+        return configProperties;
+    }
+
+    @Override
+    public String getId() {
+        return PROVIDER_ID;
+    }
+
+    @Override
+    public String getDisplayType() {
+        return "Authentication Context Class Reference (ACR)";
+    }
+
+    @Override
+    public String getDisplayCategory() {
+        return TOKEN_MAPPER_CATEGORY;
+    }
+
+    @Override
+    public String getHelpText() {
+        return "Maps the achieved LoA (Level of Authentication) to the 'acr' claim of the token";
+    }
+
+    @Override
+    protected void setClaim(IDToken token, ProtocolMapperModel mappingModel, UserSessionModel userSession, KeycloakSession keycloakSession,
+                            ClientSessionContext clientSessionCtx) {
+        AuthenticatedClientSessionModel clientSession = clientSessionCtx.getClientSession();
+        String acr = getAcr(clientSession);
+        token.setAcr(acr);
+    }
+
+    public static ProtocolMapperModel create(String name, boolean accessToken, boolean idToken) {
+        ProtocolMapperModel mapper = new ProtocolMapperModel();
+        mapper.setName(name);
+        mapper.setProtocolMapper(PROVIDER_ID);
+        mapper.setProtocol(OIDCLoginProtocol.LOGIN_PROTOCOL);
+        Map<String, String> config = new HashMap<>();
+        if (accessToken) config.put(OIDCAttributeMapperHelper.INCLUDE_IN_ACCESS_TOKEN, "true");
+        if (idToken) config.put(OIDCAttributeMapperHelper.INCLUDE_IN_ID_TOKEN, "true");
+        mapper.setConfig(config);
+        return mapper;
+    }
+
+    protected String getAcr(AuthenticatedClientSessionModel clientSession) {
+        int loa = LoAUtil.getCurrentLevelOfAuthentication(clientSession);
+        logger.tracef("Loa level when authenticated to client %s: %d", clientSession.getClient().getClientId(), loa);
+        if (loa < Constants.MINIMUM_LOA) {
+            loa = AuthenticationManager.isSSOAuthentication(clientSession) ? 0 : 1;
+        }
+
+        Map<String, Integer> acrLoaMap = AcrUtils.getAcrLoaMap(clientSession.getClient());
+        String acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, AcrUtils.getRequiredAcrValues(
+                clientSession.getNote(OIDCLoginProtocol.CLAIMS_PARAM)));
+        if (acr == null) {
+            acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, AcrUtils.getAcrValues(
+                    clientSession.getNote(OIDCLoginProtocol.CLAIMS_PARAM),
+                    clientSession.getNote(OIDCLoginProtocol.ACR_PARAM), clientSession.getClient()));
+            if (acr == null) {
+                acr = AcrUtils.mapLoaToAcr(loa, acrLoaMap, acrLoaMap.keySet());
+                if (acr == null) {
+                    acr = String.valueOf(loa);
+                }
+            }
+        }
+
```

Comment:
TBH I am more towards to keep the current way. In case that "acr" is filled on line 110, there is only single "if" check needed instead of more checks. So for example during debugging (or reading the code), you know that you can ignore the whole code inside 112-122 due you know that "acr" is filled.

Is it ok to you to keep like in current PR?

Reply:
@mposolda Yes.

Reply:
@mposolda Agree

Reply:
@tnorimat @mabartos Thanks for the confirm